Creation of ASR api

ar13pit opened the pull request:
Closes #2 
Closes tue-robotics/tue_robocup#1014

ar13pit committed:
feat: Convert test_live into Asr class

ar13pit committed:
fix: Fix variable name queue

ar13pit committed:
docs: Add docstring to Asr class

ar13pit committed:
feat: Use a cleaner interface for makedirs

ar13pit committed:
feat: Add templating for wav_out_fmt

ar13pit committed:
test: Add test for ASR api

ar13pit committed:
feat: Add python2 support for os.makedirs(..., exist_ok=True)

ar13pit commented:
A basic framework for ASR API is ready. But somehow it seems to be slower than the script `test_live.py`.

LoyVanBeek commented on src/python/yapykaldi/asr.py:
```diff
@@ -0,0 +1,161 @@
+from __future__ import (print_function, division, absolute_import, unicode_literals)
+from builtins import *
```
What Python version are you using? This line does not work on Python 2. And I don't think You are using any imports offered via this and if you were, then I'd import them one by one specifically.

ar13pit replied:
If you have `python-future` installed then it will work. It is the standard way of futurizing a python script to be compatible with 2 and 3. We could use the specific imports, but as this PR is still a prototype and not ready be merged to master, I let this be.

LoyVanBeek replied:
Can you add that to the requirements then plz?

ar13pit replied:
Done in a new PR.

MatthijsBurgh commented on src/python/yapykaldi/asr.py:
```diff
@@ -0,0 +1,161 @@
+from __future__ import (print_function, division, absolute_import, unicode_literals)
```
Why the brackets?

ar13pit replied:
Just semantics. Useful when split over multiple lines.

ar13pit commented:
@reinzor @LoyVanBeek can you guys look into the multiprocessing stuff I have in this PR? The design isn't very good. Although I have tested it thoroughly, I have observed latency issues. I adapted this from one of the test scripts I wrote `testtest_live.py` and there I do not have any object oriented design, so haven't faced any latency issues either.

LoyVanBeek commented on src/python/yapykaldi/asr.py:
```diff
@@ -0,0 +1,161 @@
+        self._queue = multiprocessing.Queue()
+        self._finalize = False
+
+        process = multiprocessing.Process(None, self.recognize, args=())
```
I don't think you actually want a different process but rather a different thread if latency is the thing multiprocessing involves serializing objects go go to a dfifferent process but that is not needed for a different *thead*, IIRC.

ar13pit replied:
Multiprocessing is needed because of concurrency. Threads execute one at a time so reading of the audio buffer will be an issue then. Also, latency is not the issue in multiprocessing.

It has occurred due to the object oriented design. The thing that is actually required is to a manager for the shared variables, make recognize and listen into two separate process by inheriting from a multiprocessing base class. Instantiate and close all of them in the start method.

LoyVanBeek commented on src/python/yapykaldi/asr.py:
```diff
@@ -0,0 +1,161 @@
+        self.timeout = timeout
+        self.wav_out_fmt = wav_out_fmt
+
+        self._p = pyaudio.PyAudio()
```
Please a better name than _p. Below I thought it was a file that was being closed.

LoyVanBeek commented on src/python/yapykaldi/asr.py:
```diff
@@ -0,0 +1,161 @@
+        logging.info("* start listening")
+
+        frames = []
+        while not self._finalize:
```
How do you intend on detecting this eventually?

ar13pit replied:
Through multiple interfaces.
1. When a preempt call is made
2. When a the grammar traversal is stuck for whatever reason
3. Timeout after completion of grammar traversal

All of these will need to trigger the interrupt handler.

LoyVanBeek replied:
OK, should work

LoyVanBeek commented on src/python/yapykaldi/asr.py:
```diff
@@ -0,0 +1,161 @@
+
+class Asr(object):
+    """API for ASR"""
+    def __init__(self, model_dir, model_type, output_dir, format=pyaudio.paInt16, channels=1, rate=16000, chunk=1024,
```
I'd separate the audio-stream listener into a separate class (eg. called `(Py)AudioStreamer`) would work better.
It allows to hide all the parameters not really meant for the Asr itself **and** to put in an alternative implementation, that eg. read a file or receives from a ROS topic.

LoyVanBeek commented on src/python/yapykaldi/asr.py:
```diff
@@ -0,0 +1,161 @@
+            else:
+                logging.info("Recognizing chunk")
+                if self.decoder.decode(self.rate, np.array(data, dtype=np.float32), self._finalize):
+                    decoded_string, _ = self.decoder.get_decoded_string()
```
To report results outside, maybe just a callback receiving the text is enough already,

LoyVanBeek committed:
Use an Event rather than a plain bool to cimmunicate between different threads/processes

LoyVanBeek committed:
Report decoded strings back via a callback

ar13pit committed:
refactor: Remove IDE files

ar13pit committed:
feat: Ignore pycharm files

ar13pit commented on src/python/yapykaldi/asr.py:
```diff
@@ -0,0 +1,178 @@
+                if self.decoder.decode(self.rate, np.array(data, dtype=np.float32), self._finalize):
+                    decoded_string, _ = self.decoder.get_decoded_string()
+                    logging.info("** {}".format(decoded_string))
+                    for cb in self._string_recognized_callbacks:
+                        cb(decoded_string)
```
This should be done after the whole recognition process else it will be sending partially decoded strings.

LoyVanBeek replied:
Ah, OK.

LoyVanBeek commented on .gitignore:
```diff
@@ -1,3 +1,5 @@
 *.egg-info/
 *.so
+
+.idea
```
Oops, sorry

LoyVanBeek committed:
Move signal handling to test script instead of reusable class

MatthijsBurgh commented on test/test_asr.py:
```diff
@@ -1,8 +1,18 @@
 
 asr = Asr(model_dir, model_type, output_dir)
+
+def interrupt_handle( sig, frame):
```
```suggestion
def interrupt_handle(sig, frame):
```

LoyVanBeek committed:
Start a class-based input API for ASR to untie the Asr class from a particular input method

LoyVanBeek committed:
Create WaveFileStreamer class to read audio from a file

ar13pit committed:
style: Remove space in arg

Co-Authored-By: Matthijs van der Burgh <MatthijsBurgh@outlook.com>

LoyVanBeek committed:
Optionally switch between live and file audio

ar13pit committed:
feat: Use Event from multiprocessing instead of threading

LoyVanBeek committed:
Also stop stream when stopping ASR

LoyVanBeek committed:
Set up stream only in start method

LoyVanBeek committed:
Remove seemingly unnecessary multiprcessing code, synchronous code seems to just work

LoyVanBeek committed:
Compose saving of recorded audio into microphone interface

LoyVanBeek committed:
Cleanup code a bit

LoyVanBeek committed:
Move initialisation of recognizer to constructor so it's ready when we finally hit 'recognize'

LoyVanBeek committed:
Test Asr with correct usage of WaveFileStreeamer

LoyVanBeek committed:
Extend callbacks to partial and full result strings

LoyVanBeek committed:
Optionally write audio to a file

LoyVanBeek committed:
Move classes for audio input and output to different submodule

LoyVanBeek committed:
Extend interface of AudioSourceBase to better support re-opening microphones nad buffering later on

LoyVanBeek committed:
fixup! Move classes for audio input and output to different submodule

LoyVanBeek committed:
Trying to make the PyAudioMicrophoneSource  async again

LoyVanBeek committed:
Add some debug sound level bars to see what the audio is doing roughly

LoyVanBeek committed:
Add more logging and exception handling

LoyVanBeek committed:
Use specifically named logger rather than generic one

LoyVanBeek committed:
Move audio stream opening/closing to the listen-thread

LoyVanBeek committed:
Only set stop Event when not already set

LoyVanBeek commented:
This took some time but is ready for review I think. Features:
- Class based API for audio sources & sinks and Asr
- Multithreading to read audio into a buffer

ar13pit committed:
fix: Remove unnecessay module init files

ar13pit committed:
test: Futurize script

ar13pit committed:
test: Fix import orders

ar13pit committed:
test: Fix linting errors

ar13pit committed:
docs: Add docstrings to the functions

ar13pit committed:
style: Formatting fixes

ar13pit commented:
@LoyVanBeek This does work live without any issues. I'm going to clean it a bit more because the prints are pretty noisy.

ar13pit committed:
feat: Move util function to independent module

ar13pit committed:
fix: Use Event class from threading module

ar13pit committed:
feat: Add support for GMM models

ar13pit committed:
fix: Fix linter issues with log strings

ar13pit committed:
docs: Add docstrings to interface

ar13pit committed:
feat: Merge callback register functions

Add arg for partial recognition callback

ar13pit committed:
refactor: Remove unused import

ar13pit committed:
fix: Cleanup logging strings

ar13pit committed:
fix: Use timeout from method arg

ar13pit committed:
fix: Move attribute declaration to init

ar13pit committed:
fix: Read file only if object is None

ar13pit committed:
refactor: Remove unnecessary else

ar13pit committed:
feat: Reset internal variables upon clean file close

ar13pit committed:
refactor: Remove unnecessary log

ar13pit committed:
test: Update callback function names and remove prints

ar13pit committed:
refactor: Combine import statements

ar13pit committed:
docs: Add module docstrings and fix typo in comment

ar13pit committed:
fix: Check framerate of opened audio file

raise RuntimeWarning if the file is already open

ar13pit committed:
feat: Version bump

ar13pit commented:
I think just few small things are remaining before we can merge this PR.
- Moving saver from audio source to ASR. We have an api to get chunks. Let the module connecting to the source (i.e. ASR) decide whether saving is required
- Better file handling in saver so that files are not overwritten and preferably support filename formats (this will be needed with the ros wrapper)
- Add support for writing out the decoded string
- Cleanup decoded string visualization in ASR
- Cleanup the test_asr script futher as the event declared in that script is currently not doing anything

ar13pit committed:
fix: Add back the stop function in class WaveFileSource

ar13pit committed:
feat: Add logging to WaveFileSource

ar13pit committed:
test: Remove unnecessary print statements

Logging functions take care of status printing

ar13pit committed:
test: Remove repeated asr and streamer stop commands

asr.stop is already in the interrupt handler, streamer.stop is being called by asr.stop

ar13pit committed:
test: Remove unnecessary event handler

It is doing nothing here

ar13pit commented:
The following things are currently broken:
1. Using an audio file streamer for ASR. 
2. Writing audio file upon completion of recognition in python3

ar13pit committed:
feat: Rename module to io for larger inclusion

ar13pit committed:
docs: Add module docstring

ar13pit committed:
refactor: Update io module imports

ar13pit committed:
fix: Fix broken sink api in python3

ar13pit committed:
test: Add test script for io module

ar13pit committed:
test: Fix reading of audio file

ar13pit committed:
feat: Use python3 primitives and futurize module

ar13pit committed:
docs: Add class constructor docstrings

ar13pit commented:
@LoyVanBeek A new issue spotted:

When using the WaveFileSource streamer, the stream misses out on probably the last chunk. You can test this out with the audio file at `data/lsen1.wav` relative to the repo root.

ar13pit committed:
feat: Move logging config to test script

ar13pit committed:
feat: Move logging to separate module

ar13pit committed:
style: Fix pylint object inheritance warning

ar13pit committed:
feat: Add args to print decoding results to log

Set to be false by default

ar13pit committed:
style: Fix linting warnings

ar13pit committed:
test: Set log args in Asr constructor

ar13pit committed:
feat: Add volume indicator function

Useful in debugging chunk volume levels

ar13pit committed:
refactor: Merge chunk unpacking and numpy array creation

ar13pit committed:
feat: Replace log flags with 1 debug flag and add volume indicator

Always print the final decoded result to log

ar13pit committed:
test: Update debug args

ar13pit committed:
docs: Add missing docstring about arg stream

ar13pit commented:
@LoyVanBeek This ready now. Have a look before I merge this.
